File: backend/app/services/order_status.py
"""Shared order status vocabulary for real-time tracking.

Pipeline (customer notifications match these milestones):
  pending → accepted → cooking → delivered
"""

from __future__ import annotations

from enum import Enum


class LiveOrderStatus(str, Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    COOKING = "cooking"
    DELIVERED = "delivered"
    REJECTED = "rejected"


# Map leftover multi-step statuses from the old pipeline onto the new one.
LEGACY_STATUS_MAP: dict[str, str] = {
    "preparing_ingredients": LiveOrderStatus.COOKING.value,
    "quality_check": LiveOrderStatus.COOKING.value,
    "packing": LiveOrderStatus.COOKING.value,
    "ready": LiveOrderStatus.DELIVERED.value,
    "out_for_delivery": LiveOrderStatus.DELIVERED.value,
}


def canonicalize_status(status: str) -> str:
    return LEGACY_STATUS_MAP.get(status, status)


# Customer-facing timeline (excludes rejected)
TRACKING_FLOW: list[LiveOrderStatus] = [
    LiveOrderStatus.PENDING,
    LiveOrderStatus.ACCEPTED,
    LiveOrderStatus.COOKING,
    LiveOrderStatus.DELIVERED,
]

STATUS_LABELS: dict[str, str] = {
    LiveOrderStatus.PENDING.value: "Pending",
    LiveOrderStatus.ACCEPTED.value: "Accepted",
    LiveOrderStatus.COOKING.value: "Cooking",
    LiveOrderStatus.DELIVERED.value: "Done",
    LiveOrderStatus.REJECTED.value: "Rejected",
}

# Chef one-click advance path after accept
ADVANCE_FLOW: list[LiveOrderStatus] = [
    LiveOrderStatus.ACCEPTED,
    LiveOrderStatus.COOKING,
    LiveOrderStatus.DELIVERED,
]

STATUS_MESSAGES: dict[str, str] = {
    LiveOrderStatus.PENDING.value: "Order placed — waiting for the kitchen.",
    LiveOrderStatus.ACCEPTED.value: "Chef accepted your order.",
    LiveOrderStatus.COOKING.value: "Your food is now Cooking",
    LiveOrderStatus.DELIVERED.value: "Done",
    LiveOrderStatus.REJECTED.value: "Your order has been rejected.",
}


def next_status(current: str) -> str | None:
    current = canonicalize_status(current)
    try:
        idx = [s.value for s in ADVANCE_FLOW].index(current)
    except ValueError:
        return None
    if idx >= len(ADVANCE_FLOW) - 1:
        return None
    return ADVANCE_FLOW[idx + 1].value


def progress_percent(status: str) -> int:
    status = canonicalize_status(status)
    if status == LiveOrderStatus.REJECTED.value:
        return 0
    try:
        idx = [s.value for s in TRACKING_FLOW].index(status)
    except ValueError:
        return 0
    if len(TRACKING_FLOW) <= 1:
        return 100
    return int(round((idx / (len(TRACKING_FLOW) - 1)) * 100))


def status_index(status: str) -> int:
    status = canonicalize_status(status)
    try:
        return [s.value for s in TRACKING_FLOW].index(status)
    except ValueError:
        return 0

File: web/src/utils/orderStatus.js
/** Live order tracking — pending → accepted → cooking → delivered. */

export const ORDER_FLOW = ['pending', 'accepted', 'cooking', 'delivered']

export const STATUS_LABELS = {
  pending: 'Pending',
  accepted: 'Accepted',
  cooking: 'Cooking',
  delivered: 'Done',
  rejected: 'Rejected',
}

/** Leftover multi-step statuses from the old pipeline. */
const LEGACY_STATUS_MAP = {
  preparing_ingredients: 'cooking',
  quality_check: 'cooking',
  packing: 'cooking',
  ready: 'delivered',
  out_for_delivery: 'delivered',
}

export function canonicalizeStatus(status) {
  return LEGACY_STATUS_MAP[status] || status
}

export const ACTIVE_STATUSES = ['accepted', 'cooking']

export function statusLabel(status) {
  const s = canonicalizeStatus(status)
  return STATUS_LABELS[s] || s
}

export function statusIndex(status) {
  const i = ORDER_FLOW.indexOf(canonicalizeStatus(status))
  return i >= 0 ? i : 0
}

export function progressPercent(status) {
  if (canonicalizeStatus(status) === 'rejected') return 0
  const i = statusIndex(status)
  if (ORDER_FLOW.length <= 1) return 100
  return Math.round((i / (ORDER_FLOW.length - 1)) * 100)
}

export function nextStatus(status) {
  const i = ORDER_FLOW.indexOf(canonicalizeStatus(status))
  if (i < 0 || i >= ORDER_FLOW.length - 1) return null
  // Chef advances from accepted → cooking → delivered (pending needs Accept)
  if (canonicalizeStatus(status) === 'pending') return null
  return ORDER_FLOW[i + 1]
}

export function normalizeServerOrder(order) {
  if (!order) return null
  const status = canonicalizeStatus(order.status)
  return {
    id: order.id,
    status,
    items: order.items || [],
    total: order.total ?? 0,
    createdAt: order.created_at ?? order.createdAt ?? Date.now(),
    updatedAt: order.updated_at ?? order.updatedAt ?? Date.now(),
    customerName: order.customer_name ?? order.customerName ?? 'Guest',
    customerEmail: order.customer_email ?? order.customerEmail ?? '',
    itemCount: order.item_count ?? order.itemCount ?? 0,
    estimatedMinutes: order.estimated_minutes ?? order.estimatedMinutes ?? 30,
    message: order.message || notificationForStatus(status) || '',
    statusLabel: order.status_label || statusLabel(status),
    progressPercent: order.progress_percent ?? progressPercent(status),
    nextStatus: order.next_status ?? nextStatus(status),
    nextStatusLabel:
      order.next_status_label ||
      statusLabel(order.next_status || nextStatus(status) || ''),
  }
}

export function isPending(status) {
  return canonicalizeStatus(status) === 'pending'
}

export function isActivePipeline(status) {
  return ACTIVE_STATUSES.includes(canonicalizeStatus(status))
}

export function isCompleted(status) {
  return canonicalizeStatus(status) === 'delivered'
}

export function isRejected(status) {
  return canonicalizeStatus(status) === 'rejected'
}

/** Customer real-time milestones only. */
export const CUSTOMER_NOTIFY_STATUSES = ['accepted', 'cooking', 'delivered']

export const NOTIFICATION_COPY = {
  accepted: '👨‍🍳 Chef accepted your order.',
  cooking: '🔥 Your food is now Cooking',
  delivered: '🎉 Done',
  rejected: '❌ Your order has been rejected.',
}

export const NOTIFICATION_TITLE = {
  accepted: 'Chef Update',
  cooking: 'Chef Update',
  delivered: 'Chef Update',
  rejected: 'Order Update',
}

export function notificationForStatus(status) {
  const s = canonicalizeStatus(status)
  if (!CUSTOMER_NOTIFY_STATUSES.includes(s) && s !== 'rejected') return null
  return NOTIFICATION_COPY[s] || null
}

export function notificationTitleForStatus(status) {
  const s = canonicalizeStatus(status)
  return NOTIFICATION_TITLE[s] || 'Chef Update'
}

export function shouldNotifyCustomer(status) {
  const s = canonicalizeStatus(status)
  return CUSTOMER_NOTIFY_STATUSES.includes(s) || s === 'rejected'
}
